tomviz/RotateAlignWidget.cxx

```diff
@@ -0,0 +1,319 @@
+                        1.0, 1.0);
+
+  double *bounds = mapper->GetBounds();
+  vtkVector3d point;
+  point[0] = 0.5 * (bounds[0] + bounds[1]);
+  point[1] = 0.5 * (bounds[2] + bounds[3]);
+  point[2] = 0.5 * (bounds[4] + bounds[5]);
+  camera->SetFocalPoint(point.GetData());
+  point[2] += 500;
+  camera->SetPosition(point.GetData());
+  camera->SetViewUp(0.0, 1.0, 0.0);
+  camera->ParallelProjectionOn();
+  camera->SetParallelScale(0.5 * (bounds[1] - bounds[0] + 1));
+}
+}
+
+namespace tomviz
+{
+
+class RotateAlignWidget::RAWInternal
+{
+public:
+  Ui::RotateAlignWidget Ui;
+  vtkNew<vtkImageSlice> mainSlice;
+  vtkNew<vtkImageSlice> reconSlice[3];
+  vtkNew<vtkImageSliceMapper> mainSliceMapper;
+  vtkNew<vtkImageSliceMapper> reconSliceMapper[3];
+  vtkNew<vtkRenderer> mainRenderer;
+  vtkNew<vtkRenderer> reconRenderer[3];
+  vtkNew<vtkImageData> reconImage;
+  QPointer<DataSource> Source;
+  vtkNew<vtkLineSource> rotationAxis;
+  vtkNew<vtkActor> axisActor;
+
+  void setupCameras()
+  {
+    setupRenderer(this->mainRenderer.Get(), this->mainSliceMapper.Get());
+    setupRenderer(this->reconRenderer[0].Get(), this->reconSliceMapper[0].Get());
+    setupRenderer(this->reconRenderer[1].Get(), this->reconSliceMapper[1].Get());
+    setupRenderer(this->reconRenderer[2].Get(), this->reconSliceMapper[2].Get());
+  }
+
+  void setupRotationAxisLine()
+  {
+    vtkTrivialProducer *t =
+        vtkTrivialProducer::SafeDownCast(this->Source->producer()->GetClientSideObject());
+    if (!t)
+    {
+      return;
+    }
+    vtkImageData *imageData = vtkImageData::SafeDownCast(t->GetOutputDataObject(0));
+    if (imageData)
+    {
+      int extent[6];
+      imageData->GetExtent(extent);
+      double spacing[3];
+      imageData->GetSpacing(spacing);
+      int projections = extent[5] - extent[4] + 1;
+      this->Ui.projection->setMaximum(projections - 1);
+      this->Ui.rotationAxis->setRange(-projections * spacing[1] / 2.0, projections * spacing[1] / 2.0);
+      double bounds[6];
+      imageData->GetBounds(bounds);
+      double point1[3], point2[3];
+      point1[0] = bounds[0] - (bounds[1] - bounds[0]);
+      point2[0] = bounds[1] + (bounds[1] - bounds[0]);
+      point1[1] = (bounds[2] + bounds[3]) / 2.0;
+      point2[1] = (bounds[2] + bounds[3]) / 2.0;
+      point1[2] = bounds[5];
+      point2[2] = bounds[5];
+      this->rotationAxis->SetPoint1(point1);
+      this->rotationAxis->SetPoint2(point2);
+      this->rotationAxis->Update();
+      this->axisActor->GetMapper()->Update();
+    }
+  }
+
+  void moveRotationAxisLine()
+  {
+    vtkTransform *tform = vtkTransform::SafeDownCast(this->axisActor->GetUserTransform());
+    if (!tform)
+    {
+      vtkNew<vtkTransform> t;
+      t->PreMultiply();
+      tform = t.Get();
+      this->axisActor->SetUserTransform(t.Get());
+    }
+    tform->Identity();
+    tform->Translate(0, this->Ui.rotationAxis->value(), 0);
+    tform->RotateZ(this->Ui.rotationAngle->value());
```

Comment:
This is what I get for working with a volume that is a cube centered at the origin... I don't see the interesting cases.